### PR TITLE
KAFKA-4379 Followup: Avoid sending to changelog while restoring InMemoryLRUCache

### PR DESCRIPTION
**Pull request by guozhangwang**
1. Added a flag to indicate if it is restoring or not in the LRU Store; since we only have a restore callback we have to set it each time applying the change.
2. Fixed the corresponding unit test, plus some minor cleaning up.

### CODE REVIEW

**Comment by guozhangwang** [comment]
ping @norwood @dguy for reviews.

The reason that I think it is safe to remove the listener is that with either log truncation and log compaction, we can actually prove that after replaying the lru store's image will not contain any unexpected records, i.e.: any record that will show up in the restored lru will definitely be in the original lru, even with `delete` operations supported in the lru store. I'd omit the proof here but we can chat offline if you are interested.

With that, I think we then do not need to log the evicted records anymore from the changelog topic, and it is safe to just remove this piece of logic here.

**Review by dguy — COMMENTED** [review]
@guozhangwang what is the downside of this? Is it that in the case it is backed by a compacted topic that items won't ever be removed as tombstones aren't sent?

**Comment by norwood** [comment]
time for a new compaction policy based on message count

**Comment by guozhangwang** [comment]
@norwood @dguy @mjsax I thought about that too. What I concluded was that if we record the evicted record along with log compaction, we actually will not get correct restored state compared with the original lru store ("state" include the ordering of the filled records which will affect future eviction decisions after restoration).

Also I did some simple test to verify that with recorded eviction record, restored state may actually be incorrect, so admittedly the downside of ever growing changelog is worse, but that is necessary for correctness.

On the other hand, since as we have discussed, we can write the offset in checkpoint file as well for lru because of the proven property mentioned above to reduce restoration latency (this is a niche optimization for lru only so not to much work to add).

**Comment by guozhangwang** [comment]
Chatted with @norwood offline, and we decided to go with the alternative direction to still log evicted records only after restoration.

**Comment by guozhangwang** [comment]
@norwood @dguy @mjsax open for review again.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java [hunk 1]
```diff
@@ -52,44 +52,41 @@
...
             private static final long serialVersionUID = 1L;
 
             @Override
             protected boolean removeEldestEntry(Map.Entry<K, V> eldest) {
-                if (super.size() > maxCacheSize) {
-                    K key = eldest.getKey();
-                    if (listener != null) listener.apply(key, eldest.getValue());
-                    return true;
+                boolean evict = super.size() > maxCacheSize;
+                if (evict && !restoring && listener != null) {
```
Comment (norwood) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java:77]: do the updates on `restoring` from the `StateRestoreCallback` always happen in the same thread?

**Review by mjsax — APPROVED** [review]
LGTM

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java:77]: Yes.

Comment (norwood) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java:77]: 👏

**Review by dguy — APPROVED** [review]
LGTM

**Review by hachikuji — APPROVED** [review]
LGTM. I'll merge after @guozhangwang changes the title.

**Comment by guozhangwang** [comment]
local unit test passed.

**Pull request closed without merging**